### PR TITLE
addr2line: use DW_AT_low_pc/DW_AT_high_pc instead of .debug_aranges

### PR DESCRIPTION
**Pull request by philipc**
.debug_aranges seems to be poorly supported by compilers and is incomplete.

This is only a partial fix for #72. I'll file issues for the other problems. This is also very slow now, so #53 is still required.

### CODE REVIEW

**Comment by khuey** [comment]
> .debug_aranges seems to be poorly supported by compilers and is incomplete.

Well that sucks :/

File: examples/addr2line.rs [hunk 1]
```diff
@@ -111,29 +69,87 @@ fn symbolicate<Endian>(file: &object::File, matches: &getopts::Matches)
...
+    for addr in addrs {
+        find_address(debug_line, &units, addr);
+    }
+}
+
+fn find_address<Endian>(debug_line: gimli::DebugLine<Endian>, units: &[Unit], addr: u64)
+    where Endian: gimli::Endianity
+{
+    for unit in units {
+        if unit.contains_address(addr) {
```
Comment (fitzgen) [inline examples/addr2line.rs]: Possible follow up: investigate sorting units by their addresses and doing a binary search here.

**Comment by fitzgen** [comment]
LGTM.

However, for the future, I think we should consider whether we want to implement a higher level, "semantic" API on top of the raw parser/DIE traversal code. Something that has a bunch of `Cow` and can become owned if the API consumer wishes. Basically a nicer type for each DW_TAG_whatever that has (internally cacheable?) methods to get the data that type is expected to have. Something like this (total sketch without lifetimes and all that):

```rust
struct Subprogram {
    ...
}

impl Subprogram {
    pub fn new(die: DebuggingInformationEntry) -> Subprogram {
        assert!(die.tag() == DW_TAG_subprogram);
        Subprogram { ... }
    }

    pub fn name(&self) -> Option<&ffi::CStr> { ... }

    pub fn linkage_name(&self) -> Option<&ffi::CStr> { ... }

    pub fn calling_convention(&self) -> Option<DwCc> { ... }

    pub fn is_pure(&self) -> Option<bool> { ... }

    pub fn is_recursive(&self) -> Option<bool> { ... }

    // Etc...
}
```

**Comment by fitzgen** [comment]
I'm imagining an API like that making the `Unit` type here obsolete.

**Comment by philipc** [comment]
Yeah we definitely need higher level APIs. These should provide the option to cache stuff, and then provide fast access to the cache (eg we should be able to cache abbreviations and line number information for a CU, and use binary search for the line number info). We also may want to provide the option to lazily parse abbreviations. eg if all we want is the CU DIE, and the CU abbreviation is first, then no need to parse them all. I want to finish validating the low level parsing before doing that though.

**Pull request merged**